e2/src/test/scala/io/prediction/e2/evaluation/CrossValidationTest.scala

```diff
@@ -0,0 +1,124 @@
+    LabeledPoint(Label1, Array(Attribute1, Attribute2)),
+    LabeledPoint(Label2, Array(NotAttribute1, Attribute2)),
+    LabeledPoint(Label3, Array(Attribute1, NotAttribute2)),
+    LabeledPoint(Label4, Array(NotAttribute1, NotAttribute2))
+  )
+
+  val dataCount = labeledPoints.size
+  val evalKs = (1 to dataCount)
+  val emptyParams = new CrossValidationTest.EmptyEvaluationParams()
+  type Fold = (CrossValidationTest.TrainingData,
+               CrossValidationTest.EmptyEvaluationParams,
+               RDD[(CrossValidationTest.Query, CrossValidationTest.ActualResult)])
+
+  def toTestTrain(dataSplit: Fold): (Seq[LabeledPoint], Seq[LabeledPoint]) = {
+    val trainingData = dataSplit._1.labeledPoints
+    val queryActual = dataSplit._3
+    val testingData = queryActual.map { case (query, actual) =>
+      LabeledPoint(actual.label, query.features)
+    }
+    (trainingData, testingData.collect().toSeq)
+  }
+
+  def splitData(k: Int, labeledPointsRDD: RDD[LabeledPoint]): Seq[Fold] = {
+    CommonHelperFunctions.splitData[
+      LabeledPoint,
+      CrossValidationTest.TrainingData,
+      CrossValidationTest.EmptyEvaluationParams,
+      CrossValidationTest.Query,
+      CrossValidationTest.ActualResult](
+      k,
+      labeledPointsRDD,
+      emptyParams,
+      CrossValidationTest.toTrainingData,
+      CrossValidationTest.toQuery,
+      CrossValidationTest.toActualResult)
+  }
+
+
+
+  "Fold count" should "equal evalK" in {
+    val labeledPointsRDD = sc.parallelize(labeledPoints)
+    val lengths = evalKs.map {k => splitData(k, labeledPointsRDD).length}
+    lengths should be(evalKs)
+  }
+
+
+  "Testing data size" should  "be within 1 of total / evalK" in {
+    val labeledPointsRDD = sc.parallelize(labeledPoints)
+    val splits = evalKs.map {k => k -> splitData(k, labeledPointsRDD)}
+    val diffs = splits.map { case (k, folds) =>
+      folds.map { fold => fold._3.count() - dataCount / k}
+    }
+    diffs.flatten.filter(_ > 1) should be('empty)
+    val (zeros, ones) = diffs.map({diffSequence => diffSequence.partition(_ == 0) }).unzip
+    ones.map(_.length) should be(evalKs.map {k => dataCount % k})
+  }
+
+
+
+  "Training + testing" should "equal original dataset" in {
+    val labeledPointsRDD = sc.parallelize(labeledPoints)
+    val splits = evalKs.map {k => splitData(k, labeledPointsRDD)}
+    val reJoined = splits.flatMap {folds =>
+      folds.map {fold =>
+        val (training, testing) = toTestTrain(fold)
+        (training ++ testing).toSet
+      }
+    }
+    reJoined should be(reJoined.map {k => labeledPoints.toSet})
+
+  }
+
+  "Training and testing" should "be disjoint" in {
+    val labeledPointsRDD = sc.parallelize(labeledPoints)
+    val splits = evalKs.map {k => k -> splitData(k, labeledPointsRDD)}
+    val intersections = splits.flatMap { case (k, folds) =>
+      folds.flatMap { fold =>
+        val (training, testing) = toTestTrain(fold)
+        training.toSet.intersect(testing.toSet)
```

Comment:
You should always overload hashCode when you overload equals, so I'm not sure this will work.

http://www.amazon.com/Effective-Java-Edition-Joshua-Bloch/dp/0321356683 page 45